server/src/main/java/org/candlepin/model/Consumer.java

```diff
@@ -893,13 +915,71 @@ public void setContentAccessMode(String contentAccessMode) {
 
     public void setAddOns(Set<String> addOns) {
         this.addOns = addOns;
+        this.updateRHCloudProfileModified();
     }
 
     public void addAddOn(String addOn) {
         this.addOns.add(addOn);
+        this.updateRHCloudProfileModified();
     }
 
     public void removeAddOn(String addOn) {
         this.addOns.remove(addOn);
+        this.updateRHCloudProfileModified();
+    }
+
+    public Date getRHCloudProfileModified() {
+        return this.rhCloudProfileModified;
+    }
+
+    public void setRHCloudProfileModified(Date rhCloudProfileModified) {
+        this.rhCloudProfileModified = rhCloudProfileModified;
+    }
+
+    public void updateRHCloudProfileModified() {
+        this.rhCloudProfileModified = new Date();
+    }
+
+    /**
+     * Check if the consumers facts have changed and contains the cloud profile facts.
+     * It returns true if incoming facts are cloud profile facts
+     *
+     * @param incomingFacts incoming facts
+     * @return a boolean
+     */
+    public boolean checkForCloudProfileFacts(Map<String, String> incomingFacts) {
+
+        if (incomingFacts == null) {
+            return false;
+        }
+
+        if (this.facts != null && this.facts.equals(incomingFacts)) {
+            return false;
+        }
+
+        for (CloudProfileFacts profileFact : CloudProfileFacts.values()) {
+
+            if (incomingFacts.containsKey(profileFact.getFact())) {
+
+                Map<String, String> filteredIncomingFact = incomingFacts.entrySet().stream()
+                    .filter(key -> key.getKey().equals(profileFact.getFact()))
+                    .collect(HashMap::new, (m, v)->m.put(v.getKey(), v.getValue()), HashMap::putAll);
+
+                for (Entry<String, String> entry : filteredIncomingFact.entrySet()) {
```

Comment:
I'm not sure why either the filtering (which will still do a 200+ iterations loop) or this second loop is needed?  At this point we already know the key we are interested is a cloud-profile fact and is contained in the incoming facts, and only need to check the Consumer's side.

Also, the whole purpose of using a HashMap (or HashSet), is that you get O(1) retrieval time, so you only need to call `containsKey` to check if a key is there, or `.get('key')` to retrieve the value. Looping over a HashMap is almost never needed. 

Thus, this whole method could be simplified to this:
```
    public boolean checkForCloudProfileFacts(Map<String, String> incomingFacts) {
        if (incomingFacts == null) {
            return false;
        }

        if (this.facts != null && this.facts.equals(incomingFacts)) {
            return false;
        }

        for (CloudProfileFacts profileFact : CloudProfileFacts.values()) {
            if (incomingFacts.containsKey(profileFact.getFact())) {
                if (this.facts == null ||
                    !this.facts.containsKey(profileFact.getFact()) ||
                    !this.facts.get(profileFact.getFact()).equals(incomingFacts.get(profileFact.getFact()))) {
                    return true;
                }
            }
        }

        return false;
    }
```

Reply:
I misunderstood the behavior of incoming facts. Thanks for explaining. 
I have corrected it as suggested.